### PR TITLE
ts: Add `skip discriminator` functionality

### PR DESCRIPTION
**Pull request by 0xShuk**
### Problem

Anchor does discriminator check when calling a method or fetching an account. This causes problem for the non-Anchor programs. There are several native programs whose types are compatible with Anchor but their IDLs won't work with Anchor due to the discriminator check. 

### Summary of Changes

- Add optional parameter, `skipDiscriminator` in the constructor function of class `Program`. When set to true, Anchor won't add 8-byte discriminator to the ix data when calling a method and won't check the first 8 bytes as discriminator when fetching a program account.

### Examples

**Example 1 - SPL Token Program**
The whole example available here: https://github.com/0xShuk/anchor-discriminator-examples/tree/main/spl-token

IDL:
```
..
"instructions": [
      {
        "name": "initializeMint2",
        "accounts": [
          {
            "name": "mint",
            "isMut": true,
            "isSigner": false
          }
        ],
        "args": [
          {
            "name": "instructionType",
            "type": {
              "defined": "TokenInstructionType"
            }
          },
          {
            "name": "decimals",
            "type": "u8"
          },
          {
            "name": "mintAuthority",
            "type": "publicKey"
          },
          {
            "name": "freezeAuthority",
            "type": {
              "option": "publicKey"
            }
          }
        ]
      }
    ],
..
```

Typescript:
```
..
let programId = new PublicKey(idl.metadata.address);
const tokenProgram = new Program<Idl>(
    idl as Idl,
    programId,
    provider, 
    undefined, 
    undefined,
    true  // skipDiscriminator is set to true here
);

// Create Account IX - System Program
const transferSOL = SystemProgram.createAccount({
    fromPubkey: keyPair.publicKey,
    newAccountPubkey: mintAccount.publicKey,
    lamports: LAMPORTS_PER_SOL * 0.0016,
    space: 82,
    programId: tokenProgram.programId
});

const tx = await tokenProgram.methods.initializeMint2(
    { initializeMint2: {} },
    6,
    keyPair.publicKey,
    keyPair.publicKey
).accounts({
    mint: mintAccount.publicKey,
})
.preInstructions([transferSOL])
.signers([mintAccount])
.rpc();

console.log("TX Successful: ", tx);
console.log("Mint Account created: ", mintAccount.publicKey.toBase58());

// Fetch the mint account
const mint = await tokenProgram.account.mint.fetch(mintAccount.publicKey.toBase58());
console.log(mint);
```

---

**Example 2 - SPL Governance Program**
The whole example available here: https://github.com/0xShuk/anchor-discriminator-examples/tree/main/spl-governance

IDL:
```
..
"instructions": [
      {
        "name": "createRealm",
        "accounts": [
          {
            "name": "governanceRealmAccount",
            "isMut": true,
            "isSigner": false
          },
          {
            "name": "realmAuthority",
            "isMut": false,
            "isSigner": false
          },
          {
            "name": "communityTokenMint",
            "isMut": false,
            "isSigner": false
          },
          {
            "name": "tokenHoldingAccount",
            "isMut": true,
            "isSigner": false
          },
          {
            "name": "payer",
            "isMut": true,
            "isSigner": true
          },
          {
            "name": "systemProgram",
            "isMut": false,
            "isSigner": false
          },
          {
            "name": "splTokenProgram",
            "isMut": false,
            "isSigner": false
          },
          {
            "name": "sysVarRent",
            "isMut": false,
            "isSigner": false
          },
          {
            "name": "councilTokenMint",
            "isMut": false,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "councilTokenHolding",
            "isMut": true,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "realmConfig",
            "isMut": true,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "communityVoterWeight",
            "isMut": false,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "maxCommunityVoterWeight",
            "isMut": false,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "councilVoterWeight",
            "isMut": false,
            "isSigner": false,
            "isOptional": true
          },
          {
            "name": "maxCouncilVoterWeight",
            "isMut": false,
            "isSigner": false,
            "isOptional": true
          }
        ],
        "args": [
          {
            "name": "instructionType",
            "type": {
              "defined": "InstructionType"
            }
          },
          {
            "name": "name",
            "type": "string"
          },
          {
            "name": "realmConfigArgs",
            "type": {
              "defined": "RealmConfigArgs"
            }
          }
        ]
      }
    ],
..
```

TypeScript:
```
..
const tx = await governanceProgram.methods.createRealm(
        { createRealm: {} },
        name, 
        {
            useCouncilMint: true,
            minCommunityWeightToCreateGovernance: new BN(1000000),
            communityTokenConfigArgs: {
                useVoterWeightAddin: false,
                useMaxVoterWeightAddin: false,
                tokenType: {
                    liquid: {}
                }
            },
            councilTokenConfigArgs: {
                useVoterWeightAddin: false,
                useMaxVoterWeightAddin: false,
                tokenType: {
                    membership: {}
                }
            },
            communityMintMaxVoterWeightSource: {
                supplyFraction: [new BN(10000000000)]
            }
        })
    .accounts({
        governanceRealmAccount,
        realmAuthority: keyPair.publicKey,
        communityTokenMint: communityMint,
        tokenHoldingAccount: communityTokenHoldingAccount,
        payer: keyPair.publicKey,
        systemProgram: SystemProgram.programId,
        splTokenProgram: token.TOKEN_PROGRAM_ID,
        sysVarRent: SYSVAR_RENT_PUBKEY,
        councilTokenMint: councilMint,
        councilTokenHolding: councilTokenHoldingAccount,
        realmConfig: realmConfig,
        communityVoterWeight: programId,
        councilVoterWeight: programId,
        maxCommunityVoterWeight: programId,
        maxCouncilVoterWeight: programId,
    })
    .rpc();

    console.log("Transaction Successful: ", tx);
    console.log("DAO Created: ", governanceRealmAccount.toBase58());

    // Fetch the DAO we just created
    const dao = await governanceProgram.account.realmV2.fetch(governanceRealmAccount);
    console.log(dao);
```

### CODE REVIEW

File: ts/packages/anchor/src/coder/borsh/accounts.ts [hunk 1]
```diff
@@ -49,7 +49,12 @@ export class BorshAccountsCoder<A extends string = string>
     return Buffer.concat([discriminator, accountData]);
   }
 
-  public decode<T = any>(accountName: A, data: Buffer): T {
+  public decode<T = any>(accountName: A, data: Buffer, skipDiscriminator?: boolean): T {
+    if (skipDiscriminator) {
+      const prepend = Buffer.from("0".repeat(16), "hex");
```
Comment (acheroncrypto) [inline ts/packages/anchor/src/coder/borsh/accounts.ts:54]: Why do we prepend data to accounts before decoding?

**Review by acheroncrypto — COMMENTED** [review]
```ts
const tx = await tokenProgram.methods.initializeMint2(
    { initializeMint2: {} },
    6,
    keyPair.publicKey,
    keyPair.publicKey
)
```

Would it be fine if we removed the first argument `{ initializeMint2: {} }` and encoded the instruction discriminator based on the method name when `skipDiscriminator` is active? It feels like unnecessary duplication.

File: ts/packages/anchor/src/coder/borsh/accounts.ts (hunk 1 above)
Comment (0xShuk) [reply ts/packages/anchor/src/coder/borsh/accounts.ts]: The `decodeUnchecked` method assumes that the data includes the discriminator. I have changed the logic now, let me know what you think.

Comment (acheroncrypto) [reply ts/packages/anchor/src/coder/borsh/accounts.ts]: Thanks, it's clearer now.

**Comment by ngundotra** [comment]
Why not just allow the discriminator to be explicitly set? `explicit_disc: [0x0, 0x3, 0x4, ... ]`

This would be more general solution and still allow for type-checking by Anchor.

**Comment by ngundotra** [comment]
It would also make transitioning from native -> anchor easier, since you could just have 2 entrypoints to each ix.

One ix with anchor discriminator, and another ix with your native ix discriminator / enum.

**Pull request closed without merging**